lib/bundler/plugin/api/source.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+require "uri"
+require "digest/sha1"
+
+module Bundler
+  module Plugin
+    class API
+      module Source
```

Comment:
Sorry, forgot to document this class.

The reason this is module is a point where I was looking for suggestions.

My justification:

Our goal here is to provide some default implementations for many of source plugin methods and these implementations should be overridable so that if something doesn't work well for a source plugin they should be able to change it. Also at the same time all the methods of `API` should also be available.

I can see this being done in two ways:

1. (The one that's implemented here) **Runtime inheritance:** We have the implementations in a `module` (since `class` can't be inherited during runtime) and include that module **when a class registers for source plugin**. The advantage is that the only class that plugin writers will have to deal with is `API` and it does all the magic.

2. **Static inheritance:** We can make `API::Source` a subclass of `API` and the source plugins will inherit `API::Source` and the others can inherit `API` only. The advantage is that there is no confusion of runtime inheritance.